### ISSUE TITLE
FutureStub hangs when wrong netty version is used

### ISSUE THREAD

**Issue by rmichela** [grpc/grpc-java#2976]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
1.3.0

### What JVM are you using (`java -version`)?
netty4.version

### What did you do?
If possible, provide a recipe for reproducing the error.

1. Accidentally use netty-codec-http2 4.1.9.Final in your pom
2. Create a future stub
3. Call `stub.someOperation().get()`

### What did you expect to see?
* The call should fail from the underlying `ClassNotFoundException`

### What did you see instead?
* A log message is printed to the console indicating the `ClassNotFoundException`
* The call to `.get()` blocks indefinitely because the underlying future never returns
* If a timeout is used when calling `.get()`, the request always times out

This problem was especially hard to diagnose because our logger is configured to exclude log messages from libraries (legacy code), so the `ClassNotFoundException` was never surfaced.

### Findings
* `ChannelExecutor.drain()` (line 89) eats all exceptions thrown when processing the outbound task queue when `DelayedClientTransport` switches from `DelayedStream` to `NettyClientStream` in `DelayedClientTransport.newStream()`.
* `GrpcHttp2HeadersDecoder.<init>` fails with a `ClassNotFoundException` when trying to load `io.netty.handler.codec.http2.internal.hpack.Decoder`.
* Since the stream switch is itself is the task that fails, `DelayedClientTransport` buffers outbound tasks indefinitely, causing the request to hang.

### Possible Mitigations
1. Propagate exceptions from `ChannelExecutor.drain()` in a "CompositeException" that aggregates all exceptions encountered. This will cause the client to explode, rather than logging and hoping someone is watching.
2. Fail conspicuously when starting gRPC with an incompatible netty version.

**Comment by carl-mastrangelo** [comment]
Some quick thoughts: 

* Pretty much all the executors in gRPC have the problem that exceptions usually go nowhere.  We don't know where to send them.  Being async means there's usually not a callstack (or, the right callstack) to unwind.  
* *Usually* our executors wrap a user provided one.  You can add an Uncaught Exception handler and provide your own executor to handle this better.
* The future being hung is unfortunate.  The executor usually runs callbacks, which are what resolve the future.  

Did your RPCs have deadlines?  That *may* be lucky enough to cancel the RPCs and not have them get wedged.  Obviously it isn't ideal, but perhaps your binary wouldn't hang.

**Comment by rmichela** [comment]
When I added a deadline or timeout to the request, it terminated after the expected duration with a timeout, however, no request was ever initiated over the wire.

I looked over the places where the `catch (Throwable t) { //log }` pattern is used with executors. In every case I saw, it looked like the log was a last ditch effort to alert the user that something that should _never_ happen had happened. Since these cases represent legitimately irrecoverable errors, perhaps _catastrophic failure_ is appropriate? gRPC can't fix the problem, and likely the user's code can't either without recompilation.

**Comment by ejona86** [comment]
Creating the actual channel is a place more likely to throw "random" exceptions (like this one, when the classpath was broken). We can probably add a try-catch to the transport initialization, maybe in `InternalSubchannel.obtainActiveTransport()` or `NettyClientTransport.start()`. We can try-catch, and then shutdown the transport. We should actually be able to fix this sort of issue well since the transport isn't ready so no RPCs have been assigned to it yet.

@zhangkun83 and @dapengzhang0 as FYI, since we may end up touching `InternalSubchannel`.

Reproduction (at v1.3.0):
```patch
diff --git a/examples/build.gradle b/examples/build.gradle
index ad065505..3b29fd48 100644
--- a/examples/build.gradle
+++ b/examples/build.gradle
@@ -25,7 +25,10 @@ repositories {
 def grpcVersion = '1.3.0' // CURRENT_GRPC_VERSION
 
 dependencies {
-  compile "io.grpc:grpc-netty:${grpcVersion}"
+  compile "io.netty:netty-all:4.1.9.Final"
+  compile ("io.grpc:grpc-netty:${grpcVersion}") {
+    exclude group: "io.netty"
+  }
   compile "io.grpc:grpc-protobuf:${grpcVersion}"
   compile "io.grpc:grpc-stub:${grpcVersion}"
 
```

```
$ ./build/install/examples/bin/hello-world-client
May 05, 2017 9:49:23 AM io.grpc.examples.helloworld.HelloWorldClient greet
INFO: Will try to greet world ...
May 05, 2017 9:49:23 AM io.grpc.internal.ChannelExecutor drain
WARNING: Runnable threw exception in ChannelExecutor
java.lang.NoClassDefFoundError: io/netty/handler/codec/http2/internal/hpack/Decoder
	at io.grpc.netty.GrpcHttp2HeadersDecoder.<init>(GrpcHttp2HeadersDecoder.java:85)
	at io.grpc.netty.GrpcHttp2HeadersDecoder$GrpcHttp2ClientHeadersDecoder.<init>(GrpcHttp2HeadersDecoder.java:147)
	at io.grpc.netty.NettyClientHandler.newHandler(NettyClientHandler.java:119)
	at io.grpc.netty.NettyClientTransport.start(NettyClientTransport.java:190)
	at io.grpc.internal.ForwardingConnectionClientTransport.start(ForwardingConnectionClientTransport.java:44)
	at io.grpc.internal.InternalSubchannel.startNewTransport(InternalSubchannel.java:216)
	at io.grpc.internal.InternalSubchannel.obtainActiveTransport(InternalSubchannel.java:186)
	at io.grpc.internal.ManagedChannelImpl$SubchannelImplImpl.obtainActiveTransport(ManagedChannelImpl.java:812)
	at io.grpc.internal.GrpcUtil.getTransportFromPickResult(GrpcUtil.java:592)
	at io.grpc.internal.DelayedClientTransport.reprocess(DelayedClientTransport.java:295)
	at io.grpc.internal.ManagedChannelImpl$LbHelperImpl$5.run(ManagedChannelImpl.java:718)
	at io.grpc.internal.ChannelExecutor.drain(ChannelExecutor.java:87)
	at io.grpc.internal.ManagedChannelImpl$LbHelperImpl.runSerialized(ManagedChannelImpl.java:709)
	at io.grpc.internal.ManagedChannelImpl$NameResolverListenerImpl.onAddresses(ManagedChannelImpl.java:758)
	at io.grpc.internal.DnsNameResolver$1.run(DnsNameResolver.java:175)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: io.netty.handler.codec.http2.internal.hpack.Decoder
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:332)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 18 more
```

**Comment by zhangkun83** [comment]
If the transport is already in a broken state, I wonder whether shutting it down guarantees propagating the error to the transport listener.

Can we say that if any exception leaked to the Channel Executor, the channel should be in an erroneous state and should not try to recover? If we can say that, we can be more aggressive, e.g., put the channel in an "error state" that sticks to an error picker and blocks any new picker updates from `LoadBalancer`.

**Comment by ejona86** [comment]
@zhangkun83, I do think we can be more aggressive in this particular case of Transport.start() failing.

**Comment by jjhart** [comment]
I ran into a similar issue with grpc v1.5.0, also due to classpath issues (an old version of closure-compiler was sneaking an old version of protobuf into the classpath, in which memoizedSize was still private).

As a result, serialization failed *after* the stub returned a future to my calling code.  Here's the serialization exception:

```
java.lang.IllegalAccessError: tried to access field com.google.protobuf.AbstractMessage.memoizedSize from class com.scalyr.grpc.QueryWorkerProto$SowParams
  at com.scalyr.grpc.QueryWorkerProto$SowParams.getSerializedSize(QueryWorkerProto.java:1229)
  at io.grpc.protobuf.lite.ProtoInputStream.available(ProtoInputStream.java:109)
  at io.grpc.internal.MessageFramer.getKnownLength(MessageFramer.java:189)
  at io.grpc.internal.MessageFramer.writePayload(MessageFramer.java:123)
  at io.grpc.internal.AbstractStream.writeMessage(AbstractStream.java:52)
  at io.grpc.internal.DelayedStream$5.run(DelayedStream.java:218)
  at io.grpc.internal.DelayedStream.drainPendingCalls(DelayedStream.java:132)
  at io.grpc.internal.DelayedStream.setStream(DelayedStream.java:101)
  at io.grpc.internal.DelayedClientTransport$PendingStream.createRealStream(DelayedClientTransport.java:351)
  at io.grpc.internal.DelayedClientTransport$PendingStream.access$200(DelayedClientTransport.java:334)
  at io.grpc.internal.DelayedClientTransport$5.run(DelayedClientTransport.java:293)
	...
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
```

The future returned by the stub was never supplied nor failed.  Note the stub used to make the call did have `withDeadlineAfter` in place; even after the timeout expired the future remain unsupplied.

I rely on the future eventually being supplied/failed in order to decrement counters used in backpressure logic.  Leaking these counters is not ideal =)

Here's my question:

Can unsupplied futures happen in other, less esoteric failure modes?  In other words, do I need to add safeguards against the future remaining empty, or do we feel pretty good that the future **will** be resolved as long as no panic-level errors occur?

Also: would moving my resource-release logic into an interceptor be more robust than using the future callbacks?  The `ClientInterceptor` doc specifies it is used to intercept calls *before* dispatch, whereas I'd need to hook into post-request-completion processing .. not sure if that's possible on the client side.


Note: to reproduce this particular failure mode without having to resort to classpath, simply insert a runtime exception into the protobuf-generated source:

```
public int getSerializedSize() {
  int size = memoizedSize;
  if (1 == 1) throw new RuntimeException("foo"); // hack to fake IllegalAccessError-type problem
```

Finally: thanks!  Great library.

**Comment by ejona86** [comment]
> In other words, do I need to add safeguards against the future remaining empty, or do we feel pretty good that the future will be resolved as long as no panic-level errors occur?

You should feel pretty good that the future will be resolved as long as no panic-level errors occur.

> would moving my resource-release logic into an interceptor be more robust than using the future callbacks?

Should be about the same.

> The ClientInterceptor doc specifies it is used to intercept calls before dispatch, whereas I'd need to hook into post-request-completion processing .. not sure if that's possible on the client side.

It's possible by making your own ClientCall.Listener (extending ForwardingClientCallListener).

**Comment by zhangkun83** [comment]
#3293 should cover this.

**Comment by zhangkun83** [comment]
This is to be resolved by #4023
